src/codegen/compile.ml

```diff
@@ -3940,13 +4012,18 @@ module Arr = struct
       get_f ^^
       (* Call *)
       Closure.call_closure env 1 1 ^^
-      store_ptr ^^
+      set_value ^^
 
-      (* Increment index *)
+      (* Recompute the array element address because the array may be moved
+         by a GC increment triggered by the tabulate function. *)
+      (* Write array element *)
+      get_r ^^
       get_i ^^
-      compile_add_const 1l ^^
-      set_i
+      idx env ^^ (* Resolves object forwarding *)
+      get_value ^^
+      store_ptr
```

Comment:
Indeed, it is the price for the allocation increments (to keep up with high allocation rates). If we would have a generational collection combined with the incremental one (with even higher complexity), we could eliminate the allocation increment and use this optimized logic. Maybe there are other possibilities for keeping this optimization?

For this PR, I believe I should consider to even remove the tabulate logic from the `compile.ml` since it is no longer optimized and could just as well be implemented in the Motoko in the compiler prelude.

Reply:
> For this PR, I believe I should consider to even remove the tabulate logic from the `compile.ml` since it is no longer optimized and could just as well be implemented in Motoko in the compiler prelude.

No, I don't think that's possible without adding some unsafe array primitives. Tabulate is used to initialize an  *immutable* array - that's why its implemented in the RTS, not motoko. Base has an implementation for mutable arrays (in Motoko) but Timo was actually asking if we could provide a quicker prim. 

This one should at least still be avoiding the bounds check on every write, so at least still a small win.

Reply:
True, I did not consider that aspect (and would have noticed during implementation). Thanks for pointing this out.

Reply:
Update: I removed the allocation increments and postponed them to next regularly scheduled GC call. With this change, we can again use the optimized tabulate function. Thanks to @ulan for this suggestion.